pkg/wireguard/wireguard.go

```diff
@@ -133,24 +136,68 @@ func (dev *Device) Run(ctx context.Context, peers PeerSource) error {
 		case <-device.Wait():
 			return nil
 		case peerEvent := <-peers.EventCh():
-			cfg := wgtypes.Config{
-				Peers: []wgtypes.PeerConfig{
-					{
-						PublicKey:         peerEvent.PubKey,
-						Remove:            peerEvent.Remove,
-						ReplaceAllowedIPs: true,
-						AllowedIPs: []net.IPNet{
-							*netaddr.IPPrefixFrom(peerEvent.Address, peerEvent.Address.BitLen()).IPNet(),
-						},
-					},
-				},
+			if err := dev.handlePeerEvent(client, logger, peerEvent); err != nil {
+				return err
 			}
+		}
+	}
+}
+
+func (dev *Device) handlePeerEvent(client *wgctrl.Client, logger *zap.Logger, peerEvent PeerEvent) error {
+	if !peerEvent.Remove {
+		oldCfg, err := client.Device(interfaceName)
+		if err != nil {
+			return fmt.Errorf("error retrieving Wireguard configuration: %w", err)
+		}
+
+		// check if this update can be skipped
+		pubKey := peerEvent.PubKey.String()
+
+		for _, oldPeer := range oldCfg.Peers {
+			if oldPeer.PublicKey.String() == pubKey {
+				if len(oldPeer.AllowedIPs) != 1 {
+					break
+				}
+
+				if prefix, ok := netaddr.FromStdIPNet(&oldPeer.AllowedIPs[0]); ok {
+					if prefix.IP() == peerEvent.Address {
+						// skip the update
+						logger.Info("skipping peer update", zap.String("public_key", pubKey))
+
+						return nil
+					}
+				}
 
-			if err = client.ConfigureDevice(interfaceName, cfg); err != nil {
-				return fmt.Errorf("error configuring Wireguard peers: %w", err)
+				break
 			}
 		}
 	}
+
+	cfg := wgtypes.Config{
+		Peers: []wgtypes.PeerConfig{
+			{
+				PublicKey: peerEvent.PubKey,
+				Remove:    peerEvent.Remove,
+			},
+		},
+	}
+
+	if !peerEvent.Remove {
```

Comment:
nit:  it seems that we could subroutinise `handleNonRemovePeerEvent()` to simplify the higher-order logic here, allowing a more readable flow, rather than two, back-to-back if statements amounting to the same thing, with the primary clauses of each being rather long.

Reply:
we could probably extract the check for no-op update which can be suppressed, but 2nd half of the method is almost same for both cases

Reply:
split the function